src/appleseed.studio/mainwindow/consolewidget.cpp

```diff
@@ -0,0 +1,119 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "consolewidget.h"
+
+// appleseed.studio headers.
+#include "outputredirector.h"
+
+// Qt headers.
+#include <QAction>
+#include <QContextMenuEvent>
+#include <QTextEdit>
+
+namespace appleseed {
+namespace studio {
+
+//
+// ConsoleWidget class implementation.
+//
+
+ConsoleWidget::ConsoleWidget(QWidget* parent)
+    : QSplitter(parent)
+{
+    output = new QTextEdit(this);
+    output->setUndoRedoEnabled(false);
+    output->setLineWrapMode(QTextEdit::WidgetWidth);
+    output->setReadOnly(true);
+    output->setTextInteractionFlags(
+        Qt::TextSelectableByMouse |
+        Qt::TextSelectableByKeyboard);
+
+    input = new QTextEdit(this);
+    input->setUndoRedoEnabled(true);
+    input->setLineWrapMode(QTextEdit::WidgetWidth);
+    input->setReadOnly(false);
+    input->setTextInteractionFlags(
+        Qt::TextSelectableByMouse |
+        Qt::TextSelectableByKeyboard |
+        Qt::TextEditable |
+        Qt::TextEditorInteraction);
+
+    this->insertWidget(0, output);
+    this->insertWidget(1, input);
+    this->setOrientation(Qt::Vertical);
+
+    m_action_execute_selection = new QAction("Execute selection as python script", this);
+    m_action_execute_selection->setShortcut(Qt::CTRL + Qt::Key_Return);
+    m_action_execute_selection->setShortcutContext(Qt::WidgetWithChildrenShortcut);
+
+    connect(m_action_execute_selection, SIGNAL(triggered()), this, SLOT(slot_execute_selection()));
+    addAction(m_action_execute_selection);
+
+    m_action_execute_all = new QAction("Execute all text in input as python script", this);
+    m_action_execute_all->setShortcut(Qt::CTRL + Qt::SHIFT + Qt::Key_Return);
+    m_action_execute_all->setShortcutContext(Qt::WidgetWithChildrenShortcut);
+
+    connect(m_action_execute_all, SIGNAL(triggered()), this, SLOT(slot_execute_all()));
+    addAction(m_action_execute_all);
+
+    m_action_focus_on_input = new QAction("Focus on console input", this);
+    m_action_focus_on_input->setShortcut(Qt::CTRL + Qt::Key_L);
+
+    connect(m_action_focus_on_input, SIGNAL(triggered()), input, SLOT(setFocus()));
+    addAction(m_action_focus_on_input);
+
+    interpreter = &PythonInterpreter::instance();
+    OutputRedirector redirector(output);
+    interpreter->redirect_output(redirector);
+}
+
+void ConsoleWidget::slot_execute_selection()
+{
+    QString selected = input->textCursor().selectedText().replace(QChar(8233), "\n");
```

Comment:
It might be better just to disable word wrapping. Python relies on indentation and us inserting 
newline characters will most often result in a syntax error in Python.